R/best_case_worst_case_emission_profile.R

```diff
@@ -0,0 +1,119 @@
+#' Calculates best case and worst case for emission profile at product level
+#'
+#' @param data Dataframe. Emissions profile product level output
+#'
+#' @return A dataframe
+#' @export
+#' @keywords internal
+#'
+#' @examples
+#' library(tiltToyData)
+#' library(readr)
+#' restore <- options(readr.show_col_types = FALSE)
+#'
+#' companies <- read_csv(toy_emissions_profile_any_companies())
+#' co2 <- read_csv(toy_emissions_profile_products_ecoinvent())
+#' europages_companies <- read_csv(toy_europages_companies())
+#' ecoinvent_activities <- read_csv(toy_ecoinvent_activities())
+#' ecoinvent_europages <- read_csv(toy_ecoinvent_europages())
+#' isic_name <- read_csv(toy_isic_name())
+#' profile_emissions_at_product_level <- profile_emissions_impl(
+#'   companies,
+#'   co2,
+#'   europages_companies = europages_companies,
+#'   ecoinvent_activities = ecoinvent_activities,
+#'   ecoinvent_europages = ecoinvent_europages,
+#'   isic = isic_name
+#' ) |>
+#'   unnest_product()
+#'
+#' result <- best_case_worst_case_emission_profile(profile_emissions_at_product_level)
+#' result
+#'
+#' # Cleanup
+#' options(restore)
+best_case_worst_case_emission_profile <- function(data) {
```

Comment:
I think your intent is to have a `best_case_worst_case()` that is polymorphic -- so we can use it in two ways:

1. Passing the `product` dataframe -- so it can be used by external users.
2. Passing the `tilt_profile` object -- so it can be used in a pipe internally in this package. 

If that's the case, then you can dramatically simplify the code using the S3 object-oriented system. It would look like this:

```r
best_case_worst_case <- function(data, ...) UseMethod("best_case_worst_case")
best_case_worst_case.data.frame <- function(data, ...) {
  # here goes the exact code you have now inside `best_case_worst_case_emission_profile()`
}
best_case_worst_case.tilt_profile <- function(data, ...) {
  # here we do somethig like
  product <- data |>
    unnest_product() |>
    # In fact we don't need the ".data.frame" but adding it here for clarity
    best_case_worst_case.data.frame()
    
  company <- unnest_company(data)
  
  nest_levels(product, company)
}
```

@kalashsinghal please confirm if this was your intent. If so, I push the code that implements polymorphism and we can go from there.

Reply:
@maurolepore We will also do best and worst case for sector profiles and transition risk score in future. This means that the function `best_case_worst_case_emission_profile` would be different for these cases but the function `best_case_worst_case.tilt_profile` will not change for these cases (as the functionality of nesting will remain same). So, I think to apply polymorphism, we can choose function name like this:

`best_case_worst_case_emission_profile.tilt_profile` and `best_case_worst_case_emission_profile.data.frame`

right? Or you think we can do this in any other way?

Reply:
@kalashsinghal 

The input of this function suggests `best_case_worst_case()` could be generalized to work with any indicator. 

``` r
devtools::load_all()
#> ℹ Loading tiltIndicatorAfter

best_case_worst_case_emission_profile_sample()
#> # A tibble: 3 × 4
#>   companies_id ep_product benchmark emission_profile
#>   <chr>        <chr>      <chr>     <chr>           
#> 1 any          one        all       low             
#> 2 any          two        all       medium          
#> 3 any          three      all       high
```

Currently the function seems to know the specific column names of emissions profile. That couples the function to that specific details of emissions profile. But it's possible to generalize it to make it truly modular. 

But that might be getting ahead of ourselves. We can focus on the issue at hand and write the simplest code possible that meets the current goal. (We can refactor when we truly need to reuse the code in a different indicator.)

I propose:

1. Don't export anything. Keep it internal to avoid the need for documentation and to maximize the ability to change names and implementations.
2. `best_case_worst_case()` is the modular function that takes the `tilt_profile` object, and passes the `product` data.frame to `best_case_worst_case_impl()` -- which does the heavy lifting.

Reply:
@maurolepore At the moment, I have not applied polymorphism and removed the functional approach used previously. I have made the export internal for the `best_case_worst_case_profile_emissions` function to better understand what the function does. However, please let me know if you think otherwise. Thanks!